Add Tensorflow landing page (#10646)
mlflow/mlflow: docs/source/deep-learning/tensorflow/guide/index.rst

Tensorflow within MLflow
=========================

In this guide we will walk you through how to use Tensorflow within MLflow. We will demonstrate
how to track your Tensorflow experiments and log your Tensorflow models to MLflow.

Autologging Tensorflow Experiments
-----------------------------------

.. attention::
    Autologging is only supported when you are using the ``model.fit()`` Keras API to train
    the model. Additionally only Tensorflow >= 2.3.0 is supported. If you are using an older version
    of Tensorflow or Tensorflow without Keras, please use manual logging.

MLflow can automatically log metrics and parameters from your Tensorflow training. To enable

> autologging, simply run :py:func:`mlflow.tensorflow.autolog()`:

Reviewer: (or just mlflow.autolog()). We might want to mention the two ways of enabling it

    with mlflow.start_run():
        model.fit(data, label, batch_size=5, epochs=2)

What is Logged by Autologging?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

By default, autologging logs the following to MLflow:

- The model summary as returned by ``model.summary()``.
- Training hyperparamers, e.g., batch size and epochs.
- Optimizer configs, e.g., optimizer name and learning rate.
- Dataset information.
- Training and validation metrics, including loss and any metrics specified in ``model.compile()``.
- Saved model after training completes in the format of TF saved model (compiled graph).

You can customize autologging behavior by passing arguments to :py:func:`mlflow.tensorflow.autolog()`,
for example if you don't want to log the dataset information, then you can run
``mlflow.tensorflow.autolog(log_dataset_info=False)``. Please refer to the API documentation
:py:func:`mlflow.tensorflow.autolog()` for full customization options.

Understanding Autologging
^^^^^^^^^^^^^^^^^^^^^^^^^

The way we autolog Tensorflow is by registering a custom callback to the Keras model via monkey patch.
Briefly we attach a MLflow callback to the Keras model that works similarly to normal Keras callbacks.

> The callback will be triggered per ``every_n_iter`` epochs to log the training metrics, and log model
> information before training, also log the saved model after training completes.

Reviewer: ```suggestion
The callback will be triggered per ``every_n_iter`` epochs to log the training metrics. At the start of training, parameters specified for the training run are logged prior to the commencement of the first training iteration. At the conclusion of training, the final tuned model is saved.
```

Logging to MLflow with Keras Callback
--------------------------------------

> As discussed in the previous section, under the hood MLflow autologging is simply using a Keras

Reviewer: ```suggestion
As discussed in the previous section, MLflow autologging for Keras models is simply using a Keras
```

> callback. You can write your custom callback to select information to log to MLflow.

Reviewer: ```suggestion
callback. If you wish to autolog additional information that isn't provided by the base autologging implementation via this default callback, you can write your own  custom callback to increase or decrease the amount of information that will be logged to MLflow.
```

Using the Predefined Callback
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

MLflow offers a predefined callback :py:class:`mlflow.tensorflow.MLflowCallback` that you can use or

> extend to log information to MLflow. The callback function provides the same functionality as autologging,

Reviewer: ```suggestion
extend to log information to MLflow. The callback function provides the same functionality as autologging 
```

        )

You can change the logging frequency in :py:class:`mlflow.tensorflow.MLflowCallback` by setting
``log_every_epoch`` and ``log_every_n_steps``, by default metrics are logged per epoch. Please refer to
the API documentation for more details.

Customizing MLflow Logging
^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can also write your own callback to log information to MLflow. To do that, you need to define
a class subclassing from `keras.callbacks.Callback <https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback>`_,
which provides hooks at various stages of training and validation, e.g., ``on_epoch_end`` and
``on_train_end`` are called separately at the end of each epoch and when the training is finished.
You can then use the callback in ``model.fit()``. Here is a simple example for logging the training metrics
in log scale:

.. code-block::

    from tensorflow import keras
    import math
    import mlflow

    class MLflowCallback(keras.callbacks.Callback):
        def on_epoch_end(self, epoch, logs=None):
            logs = logs or {}
            for k, v in logs.items():
                mlflow.log_metric(f"log_{k}", math.log(v), step=epoch)

> Briefly ``logs`` object will contain your training metrics (``loss`` and ``metrics`` as defined in ``model.compile()``),

Reviewer 2: Maybe Briefly can be changed to After each epoch for better clarity? and also I think the training and validation metrics can be reworded to be clearer?

Author: sg, done!

> Briefly ``logs`` object will contain your training metrics (``loss`` and ``metrics`` as defined in ``model.compile()``),

Reviewer: ```suggestion
At the conclusion of each epoch, the ``logs`` object will contain your training metrics (``loss`` and ``metrics`` as defined in ``model.compile()``),
```

> during training and validation metrics during validation. For full documentation of the Keras callback API, please

Reviewer: ```suggestion
that were calculated during the respective training iteration. Validation metrics are collected during the validation stage. For full documentation of the Keras callback API, please
```

read `keras.callbacks.Callback <https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback>`_.

Saving Your Tensorflow Model to MLflow
--------------------------------------

If you have turned on the autologging, your Tensorflow model will be automatically saved after the training

> is done, or you can explicitly save you model to MLflow by calling :py:func:`mlflow.tensorflow.log_model()`.

Reviewer 2: typo: save your model

> is done, or you can explicitly save you model to MLflow by calling :py:func:`mlflow.tensorflow.log_model()`.

Reviewer: ```suggestion
is done. If you would prefer to explicitly define the saving of your model, you can instead explicitly save your model to MLflow by calling :py:func:`mlflow.tensorflow.log_model()`.
```

After saving, you can load back the model using :py:func:`mlflow.tensorflow.load_model()`. The loaded model

> can be used for inference by calling ``predict()`` method.

Reviewer: ```suggestion
can be used for inference by calling the ``predict()`` method on the loaded model object.
```

.. code-block:: python

    import mlflow
    import tensorflow as tf
    from tensorflow import keras

    model = keras.Sequential(
        [
            keras.Input([28, 28, 3]),
            keras.layers.Conv2D(8, 2),
            keras.layers.MaxPool2D(2),
            keras.layers.Flatten(),
            keras.layers.Dense(2),
            keras.layers.Softmax(),
        ]
    )

    save_path = "model"
    with mlflow.start_run() as run:
        mlflow.tensorflow.log_model(model, "model")

    # Load back the model.
    loaded_model = mlflow.tensorflow.load_model(f"runs:/{run.info.run_id}/{save_path}")

    print(loaded_model.predict(tf.random.uniform([1, 28, 28, 3])))

Diving into Saving
^^^^^^^^^^^^^^^^^^

Under the hood of saving, we are converting the Tensorflow model into a pyfunc model, which is a generic
type of model in MLflow. The pyfunc model is saved to MLflow. You don't need to learn the basics of pyfunc
model to use Tensorflow flavor, but if you are interested, please refer to `MLflow pyfunc model <https://mlflow.org/docs/latest/models.html#how-to-load-and-score-python-function-models>`_.

Saving Format
~~~~~~~~~~~~~

> By default MLflow saves your Tensorflow model in the format of TF saved model (compiled graph), which is

Reviewer: ```suggestion
By default, MLflow saves your Tensorflow model in the format of a TF saved model (compiled graph), which is
```

suitable for deployment across platforms. You can also save your model in other formats, i.e., ``h5`` and
``keras`` by setting the ``keras_model_kwargs`` parameter in :py:func:`mlflow.tensorflow.log_model()`. For

> example. if you want to save your model in ``h5`` format, which only saves model weights instead of the
> compiled graph, you can run:

Reviewer: ```suggestion
example, if you want to save your model in ``h5`` format (which only saves model weights instead of the
compiled graph) you can run:
```

.. code-block:: python

    import mlflow
    import tensorflow as tf
    from tensorflow import keras

    model = keras.Sequential(
        [
            keras.Input([28, 28, 3]),
            keras.layers.Conv2D(8, 2),
            keras.layers.MaxPool2D(2),
            keras.layers.Flatten(),
            keras.layers.Dense(2),
            keras.layers.Softmax(),
        ]
    )

    save_path = "model"
    with mlflow.start_run() as run:
        mlflow.tensorflow.log_model(
            model, "model", keras_model_kwargs={"save_format": "h5"}
        )

    # Load back the model.
    loaded_model = mlflow.tensorflow.load_model(f"runs:/{run.info.run_id}/{save_path}")

    print(loaded_model.predict(tf.random.uniform([1, 28, 28, 3])))

> For differennce between the formats, please refer to `Tensorflow Save and Load Guide <https://www.tensorflow.org/guide/keras/save_and_serialize>`_.

Reviewer 2: typo: difference

> For differennce between the formats, please refer to `Tensorflow Save and Load Guide <https://www.tensorflow.org/guide/keras/save_and_serialize>`_.

Reviewer: ```suggestion
For difference between the formats, please refer to `Tensorflow Save and Load Guide <https://www.tensorflow.org/guide/keras/save_and_serialize>`_.
```

> Please note that if you want to deploy your model, then you need to save your model in the TF saved model format.

Reviewer: ```suggestion
Please note that if you want to deploy your model, you will need to save your model in the TF saved model format.
```

Model Signature
~~~~~~~~~~~~~~~

> Model signature is a description of a model's input and output. If your have enabled autologging and provided

Reviewer: ```suggestion
A Model Signature is a description of a model's input and output. If you have enabled autologging and provided
```

a dataset, then the signature will be automatically inferred from the dataset. Otherwise, you need to provide

> a signature in order to show the signature information on MLflow UI. A model signature will be shown in the MLflow

Reviewer: ```suggestion
a signature in order to have the signature information viewable within the MLflow UI. A model signature will be shown in the MLflow
```

UI as follows:

.. figure:: ../../../_static/images/deep-learning/tensorflow/guide/tensorflow-model-signature.png
   :alt: Tensorflow Model Signature
   :width: 90%
   :align: center

> To manually set up the signature, you can pass a ``signature`` parameter to :py:func:`mlflow.tensorflow.log_model()`.

Reviewer: ```suggestion
To manually set the signature for your model, you can pass a ``signature`` parameter to :py:func:`mlflow.tensorflow.log_model()`.
```

You will need to set the input schema by specifying the ``dtype`` and ``shape`` of the input tensors, and wrap
it with :py:func:`mlflow.types.TensorSpec`. For example,

.. code-block::

    import mlflow
    import tensorflow as tf
    import numpy as np

    from tensorflow import keras
    from mlflow.types import Schema, TensorSpec
    from mlflow.models import ModelSignature

    model = keras.Sequential([
        keras.Input([28, 28, 3]),
        keras.layers.Conv2D(8, 2),
        keras.layers.MaxPool2D(2),
        keras.layers.Flatten(),
        keras.layers.Dense(2),
        keras.layers.Softmax(),
    ])

    input_schema = Schema(
        [
            TensorSpec(np.dtype(np.float32), (-1, 28, 28, 3), "input"),
        ]
    )
    signature = ModelSignature(inputs=input_schema)

    with mlflow.start_run() as run:
        mlflow.tensorflow.log_model(model, "model", signature=signature)

    # Load back the model.
    loaded_model = mlflow.tensorflow.load_model(f"runs:/{run.info.run_id}/{save_path}")

    print(loaded_model.predict(tf.random.uniform([1, 28, 28, 3])))

Please note that model signature is not necessary for loading the model, you can still load the model

> and run inferenece if you know the input format. However, it's a good practice to include the signature

Reviewer 2: typo: inference

> Please note that model signature is not necessary for loading the model, you can still load the model
> and run inferenece if you know the input format. However, it's a good practice to include the signature
> for better model understanding.

Reviewer: ```suggestion
Please note that a model signature is not necessary for loading a model. You can still load the model 
and perform inferenece if you know the input format. However, it's a good practice to include the signature
for better model understanding.
```
